metagraph/src/graph/alignment/aligner_aggregator.hpp

```diff
@@ -4,258 +4,229 @@
-          : query_(query), rc_query_(rc_query), config_(config), graph_(graph) {
+    static constexpr Column ncol = std::numeric_limits<Column>::max();
+
+    explicit AlignmentAggregator(const DBGAlignerConfig &config) : config_(config) {
         assert(config_.num_alternative_paths);
     }
 
-    void add_alignment(Alignment&& alignment);
-
-    score_t get_min_path_score() const;
-    score_t get_max_path_score() const;
+    bool add_alignment(Alignment&& alignment);
 
-    score_t get_min_path_score(const Alignment &) const { return get_min_path_score(); }
-    score_t get_max_path_score(const Alignment &) const { return get_max_path_score(); }
+    score_t get_min_path_score(const Vector<Column> &labels) const;
+    score_t get_max_path_score(const Vector<Column> &labels) const;
 
-    const Alignment& maximum() const { return path_queue_.maximum(); }
-    void pop_maximum() { path_queue_.pop_maximum(); }
+    score_t get_min_path_score(Column label = ncol) const;
+    score_t get_max_path_score(Column label = ncol) const;
 
     std::vector<Alignment> get_alignments();
 
-    size_t size() const { return path_queue_.size(); }
-
-    bool empty() const { return path_queue_.empty(); }
+    const VectorMap<Column, PathQueue>& data() const { return path_queue_; }
+    VectorMap<Column, PathQueue>& data() { return path_queue_; }
 
     void clear() { path_queue_.clear(); }
 
-    std::string_view get_query(bool is_reverse_complement) const {
-        return is_reverse_complement ? rc_query_ : query_;
-    }
-
   private:
-    std::string_view query_;
-    std::string_view rc_query_;
     const DBGAlignerConfig &config_;
-    const DeBruijnGraph &graph_;
-    PathQueue path_queue_;
-    AlignmentCompare cmp_;
-
-    void construct_alignment_chains();
-    void construct_alignment_chain(std::string_view query,
-                                   Alignment&& chain,
-                                   typename std::vector<Alignment>::iterator begin,
-                                   typename std::vector<Alignment>::iterator end,
-                                   std::vector<score_t> &best_score,
-                                   const std::function<void(Alignment&&)> &callback);
+    VectorMap<Column, PathQueue> path_queue_;
+    ValCmp cmp_;
 };
 
 
 template <class AlignmentCompare>
-inline void AlignmentAggregator<AlignmentCompare>::add_alignment(Alignment&& alignment) {
-    if (std::find(path_queue_.begin(), path_queue_.end(), alignment) != path_queue_.end())
-        return;
-
-    if (config_.chain_alignments || path_queue_.size() < config_.num_alternative_paths) {
-        path_queue_.emplace(std::move(alignment));
-    } else if (!cmp_(alignment, path_queue_.minimum())) {
-        path_queue_.update(path_queue_.begin(), std::move(alignment));
+inline bool AlignmentAggregator<AlignmentCompare>::add_alignment(Alignment&& alignment) {
+    auto a = std::make_shared<Alignment>(std::move(alignment));
```

Comment:
Please double-check this function. I didn't change the logic but did some cleanups, so now it's easier to spot issues.
It seems strange to me. For instance, sometimes you check for the presence of an equal alignment, and sometimes you don't.

Reply:
Can you also add some comments to explain what is going on on each step? I'm trying to reverse-engineer the logic, but I think it has bugs if I'm understanding it correctly.